File: src/main/java/org/c3lang/intellij/psi/impl/C3ParserUtil.java
package org.c3lang.intellij.psi.impl;

import com.intellij.lang.PsiBuilder;
import com.intellij.lang.parser.GeneratedParserUtilBase;
import com.intellij.psi.tree.TokenSet;
import org.c3lang.intellij.psi.C3Types;

public class C3ParserUtil extends GeneratedParserUtilBase
{
	private static final TokenSet OTHER_IDENT_TYPES = TokenSet.create(
		C3Types.CONST_IDENT, C3Types.AT_IDENT, C3Types.TYPE_IDENT, C3Types.AT_TYPE_IDENT
	);

	public static boolean parsePathIdent(PsiBuilder b, int level)
	{
		PsiBuilder.Marker marker = b.mark();

		if (consumeToken(b, C3Types.CT_IDENT) || consumeToken(b, C3Types.HASH_IDENT))
		{
			marker.drop();
			return true;
		}

		if (!consumeToken(b, C3Types.IDENT))
		{
			marker.rollbackTo();
			return false;
		}

		boolean scope = b.getTokenType() == C3Types.SCOPE;
		while (consumeToken(b, C3Types.SCOPE))
		{
			if (b.lookAhead(1) != C3Types.SCOPE)
			{
				break;
			}
			consumeToken(b, C3Types.IDENT);
		}

		if (OTHER_IDENT_TYPES.contains(b.getTokenType()))
		{
			marker.rollbackTo();
			return false;
		}

		if (scope)
		{
			marker.done(C3Types.PATH);
		}
		else
		{
			marker.drop();
		}

		if (b.getTokenType() == C3Types.IDENT)
		{
			consumeToken(b, C3Types.IDENT);
		}

		return true;
	}
}

File: src/main/java/org/c3lang/intellij/psi/impl/C3ParserUtil.kt
package org.c3lang.intellij.psi.impl

import com.intellij.lang.PsiBuilder
import com.intellij.lang.parser.GeneratedParserUtilBase
import com.intellij.psi.TokenType
import com.intellij.psi.tree.TokenSet
import org.c3lang.intellij.psi.C3Types.*

object C3ParserUtil : GeneratedParserUtilBase() {
    private val otherIdentTypes = TokenSet.create(
        CONST_IDENT, AT_IDENT, TYPE_IDENT, AT_TYPE_IDENT
    )

    @JvmStatic
    fun parsePathIdent(
        b: PsiBuilder,
        level: Int
    ): Boolean {
        val marker: PsiBuilder.Marker = b.mark()

        if (!consumeToken(b, IDENT)) {
            marker.rollbackTo()
            return false
        }

        val scope = b.tokenType == SCOPE
        while (consumeToken(b, SCOPE)) {
            if (b.lookAhead(1) != SCOPE) {
                break
            }
            consumeToken(b, IDENT)
        }

        if (otherIdentTypes.contains(b.tokenType)) {
            marker.rollbackTo()
            return false
        }

        if (scope) {
            marker.done(PATH)
        } else {
            marker.drop()
        }

        if (b.tokenType == IDENT) {
            consumeToken(b, IDENT)
        }

        return true
    }

//    @JvmStatic
//    fun parsePathIdent(
//        b: PsiBuilder,
//        level: Int
//    ): Boolean {
//        val marker: PsiBuilder.Marker = b.mark()
//
//        if (!consumeToken(b, IDENT)) {
//            marker.rollbackTo()
//            return false
//        }
//
//        while (consumeToken(b, SCOPE)) {
//            if (b.lookAhead(1) != SCOPE) {
//                break
//            }
//            if (!consumeToken(b, IDENT)) {
//                // any other legal next token with "path"
//                when (b.tokenType) {
//                    CONST_IDENT, AT_IDENT, TYPE_IDENT, AT_TYPE_IDENT -> {
//                        marker.rollbackTo()
//                        return false
//                    }
//
//                    else -> {
//                        marker.done(PATH)
//                        return true
//                    }
//                }
//            }
//        }
//
//        marker.done(PATH)
//
//        if (b.tokenType == IDENT) {
//            consumeToken(b, IDENT)
//        }
//
//        return true
//    }
}